### ISSUE TITLE
Caching already hashed files during one build

### ISSUE THREAD

**Issue by pavsa** [ccache/ccache#377]
I was thinking about caching already hashed files in the depend mode for one build. I have noticed the file hashes for dependencies are cached in one process, but not across processes. 

I have some project that has under around 20k sources, each with 100-1000 dependencies and a lot of those dependencies are not unique and files themselves do not change during the build.

Lets assume the average number of deps/source is 250 and one dependency has 2kB. 
So, each time we do a full build, we have to hash 250deps * 2kB * 20000 sources=
9.5GBs of data for just the dependencies. 
Now lets assume total the number of unique dependencies is 30k. If we cached the hashes of the dependencies, we would only have to do hashing of 30000*2kB = 60MB data.

That's 162 times less data to hash.

There's of course some overhead of syncing the cache across processes and some other problems.

Is this feasible?

**Comment by afbjorklund** [comment]
I experimented a bit with keeping a memcached of the hashes (will see if I can find the branch...)

**Comment by afbjorklund** [comment]
Here it was: c8f3438538cbb1040c1fca395ba4c02cb2f6f16a

**Comment by pavsa** [comment]
Interesting, what were the results? was it faster?

**Comment by afbjorklund** [comment]
It was faster than reading the files, but then again you could just use modification dates in the first place. So it was mostly useful when you wanted to use stat only, but still needed the hash for some other reason ?

**Comment by afbjorklund** [comment]
Should be clear from the code, but the idea was to cache the stat and the hash for each absolute path. Then next time, if the stat still matched you would return the cached hash rather than recalculating it again.

**Comment by jrosdahl** [comment]
>  files themselves do not change during the build

Sounds like your use case could be solved by `sloppiness = file_stat_matches`? Have you tried that option?

**Comment by pavsa** [comment]
It won’t work in the following workflow we use:
Multiple developers share one cache and I want to reuse what others have compiled after I do my git pull.

**Comment by jrosdahl** [comment]
> Multiple developers share one cache and I want to reuse what others have compiled after I do my git pull.

I see. But when you have done your Git pull, how can you then detect that you have source code that is the same as somebody elses if you neither hash the files nor check their stat info? Ask Git about it?

**Comment by afbjorklund** [comment]
I think they do want to check the hashes, but only want to do it once per build. Assuming that local files aren't changed (without modifying time stamp / stat), it should be possible to cache their hash locally...

**Comment by afbjorklund** [comment]
Did some experiments* with resetting the modification times to match git, that should avoid having to hash even the first time. But did require some preparations, and then setting that extra sloppiness flag. (* #286)

**Comment by pavsa** [comment]
> > Multiple developers share one cache and I want to reuse what others have compiled after I do my git pull.
> 
> I see. But when you have done your Git pull, how can you then detect that you have source code that is the same as somebody elses if you neither hash the files nor check their stat info? Ask Git about it?
It would work the same as it does now, everything is checked by hashing.

But the difference would be:  not recomputing hash for the same file more than once. The point is to have a volatile hash cache that is only valid for a *single* build:

Implemented as a hash table with key = absolute path to file and value being the hash of the contents of the file. 
This cache is cleaned when you trigger build “make”.  Then, as files are being compiled and ccache is being executed, files need to be hashed by ccache, so ccache queries this hash cache if it has the hash pre-calculated for a file it needs the hash for. If it has a hash in the hash table, it returns the cached hash, if it does not, it hashes the file, puts it in the table and returns it.
 
We also use compiler check set to content, so it has to hash the compiler binary for every invokation. Having it hashed only once per build would save something too.

**Comment by jrosdahl** [comment]
I still don't understand.

Let's call the volatile cache we're talking about a _path-based cache_ (only path is the key), in contrast to the standard _content-based cache_ and less standard _stat-based cache_ ( `sloppiness = file_stat_matches`).

First you said "multiple developers share one cache and I want to reuse what others have compiled after I do my git pull". But then:

> This cache is cleaned when you trigger build “make”. Then, as files are being compiled and ccache is being executed, files need to be hashed by ccache, so ccache queries this hash cache if it has the hash pre-calculated for a file it needs the hash for. If it has a hash in the hash table, it returns the cached hash, if it does not, it hashes the file, puts it in the table and returns it.

If the path-based cache is cleaned when starting your build, then you won't be able to reuse what others have _hashed_, right? OK, so you build up a path-based cache during the build, so that a given include file is only hashed once during the whole build. Do I understand you correctly so far? If so, this just sounds like an inferior version of the stat-based cache to me:

* The stat-based cache (or parts of it) will be validated on `git pull`, not when starting the build.
* Just like the path-based cache, the stat-based cache will also only hash a given include file once during the whole build (except of course if the header file is changed during the build, in which case the file will be rehashed unlike in the path-based cache case).

If I have missed something, please explain your idea in more detail.

> We also use compiler check set to content, so it has to hash the compiler binary for every invokation. Having it hashed only once per build would save something too.

That's easy to improve: Just precompute the compiler binary hash, or any other string that represents the compiler version, and set `CCACHE_COMPILERCHECK=string:$VALUE` when building.

Also, maybe I should comment on this as well:

>  I have noticed the file hashes for dependencies are cached in one process, but not across processes.

That's safe to do since include files can be assumed not to change during a simple compilation (and if they do, it's arbitrary what happens anyway). That assumption doesn't hold for consecutive compilations since the header file may have been modified inbetween.

**Comment by pavsa** [comment]
> 
> 
> I still don't understand.
> 
> Let's call the volatile cache we're talking about a _path-based cache_ (only path is the key), in contrast to the standard _content-based cache_ and less standard _stat-based cache_ ( `sloppiness = file_stat_matches`).
> 
> First you said "multiple developers share one cache and I want to reuse what others have compiled after I do my git pull". But then:
> 
> > This cache is cleaned when you trigger build “make”. Then, as files are being compiled and ccache is being executed, files need to be hashed by ccache, so ccache queries this hash cache if it has the hash pre-calculated for a file it needs the hash for. If it has a hash in the hash table, it returns the cached hash, if it does not, it hashes the file, puts it in the table and returns it.
> 
> If the path-based cache is cleaned when starting your build, then you won't be able to reuse what others have _hashed_, right? OK, so you build up a path-based cache during the build, so that a given include file is only hashed once during the whole build. Do I understand you correctly so far? If so, this just sounds like an inferior version of the stat-based cache to me:
> 
>     * The stat-based cache (or parts of it) will be validated on `git pull`, not when starting the build.
> 
>     * Just like the path-based cache, the stat-based cache will also only hash a given include file once during the whole build (except of course if the header file is changed during the build, in which case the file will be rehashed unlike in the path-based cache case).
> 
> 
> If I have missed something, please explain your idea in more detail.

I understand what you mean now. It makes sense for a single machine and single user and single code location.
But where we work, we have multiple directories on the same machine where we have the same repository cloned (that's the workflow I have to work with). 
Assume that I have two directories where the exact same code resides: location1 and location2 do the following with stat-based cache:
1. I do `git pull` in location1 and execute `make`. No cache hits for the new commits, obvious reasons.
2. Then I do `git pull` in location2 and execute `make`. Now, the ctime&mtime will be different as git pull was executed in a different point in time in location2 then was in location1. So, everything will still have to hashed, as stat will not match.
 
Now that I think about it, the stat based cache is still inefficient, since it still hashes the same include file every time it is included in source files when saving to cache. If you include one header in 1000 source files, it will still get hashed 1000 times if the cache is empty.
You only skip the hashing in case you get the cache hit, but not during saving it to manifest.

This is the prime motivation for me to cache the hashes during one build. 

> 
> > We also use compiler check set to content, so it has to hash the compiler binary for every invokation. Having it hashed only once per build would save something too.
> 
> That's easy to improve: Just precompute the compiler binary hash, or any other string that represents the compiler version, and set `CCACHE_COMPILERCHECK=string:$VALUE` when building.
> 
Easy if you have one compiler, but we cross-compile to multiple platforms with multiple compilers in one build. There would have to be a different value for each compiler. You can't do that in single ccache config. Now it's done automatically by compiler's content. Having hash cache would save time by  hashing compiler only once.

> Also, maybe I should comment on this as well:
> 
> > I have noticed the file hashes for dependencies are cached in one process, but not across processes.
> 
> That's safe to do since include files can be assumed not to change during a simple compilation (and if they do, it's arbitrary what happens anyway). That assumption doesn't hold for consecutive compilations since the header file may have been modified inbetween.

**Comment by afbjorklund** [comment]
> > I have noticed the file hashes for dependencies are cached in one process, but not across processes.
> 
> That's safe to do since include files can be assumed not to change during a simple compilation (and if they do, it's arbitrary what happens anyway). That assumption doesn't hold for consecutive compilations since the header file may have been modified inbetween.

The theory here is that it might be safe to extend this assumption to one full build (e.g. `make`), and just not only to one compilation (`cc`). At least not change file contents without also modifying the file timestamp.

**Comment by afbjorklund** [comment]
> 2\. Now, the ctime&mtime saved in the manifests will be different for the new commits compared to times in the location2, even though they have exactly the same content, as git pull was executed in a different point in time in location2 than it was in location1.

Did you get a chance to play with using `git-restore-mtime` yet ?

**Comment by afbjorklund** [comment]
Should also note that some build systems keep their _compiler_ in a git repository too, which means that the issue above is always true. The current workaround there is to use `CCACHE_COMPILERCHECK=content`

**Comment by pavsa** [comment]
> 
> 
> > > I have noticed the file hashes for dependencies are cached in one process, but not across processes.
> > 
> > 
> > That's safe to do since include files can be assumed not to change during a simple compilation (and if they do, it's arbitrary what happens anyway). That assumption doesn't hold for consecutive compilations since the header file may have been modified inbetween.
> 
> The theory here is that it might be safe to extend this assumption to one full build (e.g. `make`), and just not only to one compilation (`cc`). At least not change file contents without also modifying the file timestamp.

You can also include stat along with the hash and check for change every time hash for that file is requested. Should be cheaper than reading+hashing.

**Comment by jrosdahl** [comment]
> Easy if you have one compiler, but we cross-compile to multiple platforms with multiple compilers in one build. There would have to be a different value for each compiler. You can't do that in single ccache config.

Right, I didn't mean that you should set it as a single config for several compilers; each unique compiler of course needs its own value. So when you have several compilers you then would have to create wrappers that set the value correctly, e.g. like this:

```
#!/bin/sh
CCACHE_COMPILERCHECK=string:something exec ccache /path/to/real/compiler "$@"
```

Or maybe play with using the _a command string_ option for compiler_check/CCACHE_COMPILERCHECK?

**Comment by jrosdahl** [comment]
> [...] This is the prime motivation for me to cache the hashes during one build.

Thanks for the clarification.

**Comment by jrosdahl** [comment]
[I'm continuing the discussion here since I think that we need to discuss how the path-based cache ("hash cache") should work before discussing its potential impact on the storage API.]

@pavsa wrote in https://github.com/ccache/ccache/issues/414#issuecomment-529230012:
> now it also makes sense to use the same API for this purpose.

Given my current interpretation of how things would work, I still disagree:

* It makes no sense to send local volatile information down to the regular storage backend API since that information then can end up being stored in a cache shared by many systems but only usable by one of them.
* It also doesn't make sense to me to store that information inside the regular ccache directory since the hash cache essentially is tied to a certain *build directory* (e.g. the Git work tree). To me it sounds natural to store that information in *that directory* instead.
* Another slightly related thing is that cache hits/misses in the hash cache is orthogonal to those of regular results, so statistics counters for normal cache usage shouldn't be affected by lookups in the hash cache.

**Comment by jrosdahl** [comment]
Implemented in #577.
